org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java

```diff
@@ -56,11 +58,23 @@ public List<Either<Command, CodeAction>> getCodeActions(Options options) {
 
 		List<Either<Command, CodeAction>> result = new ArrayList<>();
 		for (Diagnostic diagnostic : options.getCodeActionParams().getContext().getDiagnostics()) {
-			List<DiagnosticResolution> resolutions = quickfixes.getResolutions(options, diagnostic).stream()
-					.sorted(Comparator.nullsLast(Comparator.comparing(DiagnosticResolution::getLabel)))
-					.collect(Collectors.toList());
-			for (DiagnosticResolution resolution : resolutions) {
-				result.add(Either.forRight(createFix(resolution, diagnostic)));
+			if (quickfixes.handlesDiagnostic(diagnostic)) {
+				try {
+					result.addAll(options.getLanguageServerAccess()
+							.doRead(options.getURI(), (ILanguageServerAccess.Context context) -> {
+								options.setDocument(context.getDocument());
+								options.setResource(context.getResource());
+
+								List<Either<Command, CodeAction>> codeActions = new ArrayList<>();
+								quickfixes.getResolutions(options, diagnostic).stream()
+										.sorted(Comparator
+												.nullsLast(Comparator.comparing(DiagnosticResolution::getLabel)))
+										.forEach(r -> codeActions.add(Either.forRight(createFix(r, diagnostic))));
+								return codeActions;
+							}).get());
+				} catch (InterruptedException | ExecutionException e) {
```

Comment:
I understand why it was done like this, but having the blocking call here looks wrong to me.
The approach might need some more thought.

Reply:
we could avoid the blocking call if we can expose the workspace manage in the ILanguageServerAccess. What about that?

Reply:
I have added a new method doSyncRead which fixes the problem, I hope you like the solution.